File: cpp/weblog.cpp
std::ostream& operator<<(std::ostream& os, const Weblog& w) {
    char buffer [1024];
    sprintf (buffer, "%-20s %-30s %-40s", w.ip_addr.c_str(), w.date.c_str(), w.request.c_str());
    return os << buffer;
}

std::istream& operator>>(std::istream& is, Weblog& ww) {
    std::string ip, dt, rq, co, sz, rf, ag;
    is >> ip >> dt >> rq >> co >> sz >> rf >> ag;
    ww.ip_addr = ip;
    ww.date = dt;
    ww.request = rq;
    ww.code = atoi(co.c_str());
    ww.size = atoi(sz.c_str());
    ww.referer = rf;
    ww.agent = ag;

    return is;
}

bool operator==(const Weblog& a, const Weblog& b) {
    return a.ip_addr==b.ip_addr && a.date==b.date;
}

bool operator!=(const Weblog& a, const Weblog& b) {
    return !(a==b);
}

File: ruby/weblog.rb
# Weblog.rb
# Define a class for holding a weblog object
# @Author FoxBoxPDX
class Weblog
  # Declare attrs
  attr_accessor :ipaddr
  attr_accessor :date
  attr_accessor :request
  attr_accessor :code
  attr_accessor :size
  attr_accessor :referer
  attr_accessor :agent

  # Initialization
  def initialize(params)
    @ipaddr  = params[:ipaddr]
    @date    = params[:date]
    @request = params[:request]
    @code    = params[:code]
    @size    = params[:size]
    @referer = params[:referer]
    @agent   = params[:agent]
  end

  # Define a way to print this thing
  def to_s
    Kernel.sprintf("%-20s %-30s %-40s", @ipaddr, @date, @request)
  end

  # Hurt me beat me treat me like a String
  def to_str
    to_s
  end
end
